parcels/include/parcels.h

```diff
@@ -459,131 +461,82 @@ static inline StatusCode temporal_interpolation_structured_grid(type_coord x, ty
-      if ((interp_method == CGRID_VELOCITY) || (interp_method == BGRID_W_VELOCITY)){ // interpolate w
-        if (gridindexingtype == NEMO){
-          xsi = 1;
-          eta = 1;
-        }
-        else if (gridindexingtype == MITGCM){
-          xsi = 0;
-          eta = 0;
-        }
-        if (grid->zdim==1)
-          return ERROR;
-      }
-      else if (interp_method == BGRID_VELOCITY){
-        zeta = 0;
-      }    
-      if (grid->zdim==1){
-        status = spatial_interpolation_bilinear(xsi, eta, data2D[0], value); CHECKSTATUS(status);
-      }
-      else{
-        status = spatial_interpolation_trilinear(xsi, eta, zeta, data3D[0], value); CHECKSTATUS(status);
-      }
-    }
-    else if (interp_method == NEAREST){
-      if (grid->zdim==1){
-        status = spatial_interpolation_nearest2D(xsi, eta, data2D[0], value); CHECKSTATUS(status);
-      }
-      else {
-        status = spatial_interpolation_nearest3D(xsi, eta, zeta, data3D[0], value); CHECKSTATUS(status);
-      }
-    }
-    else if ((interp_method == CGRID_TRACER) || (interp_method == BGRID_TRACER)){
-      if (grid->zdim==1){
-        status = spatial_interpolation_tracer_c_grid_2D(data2D[0], value); CHECKSTATUS(status);
-      }
-      else {
-        status = spatial_interpolation_tracer_c_grid_3D(data3D[0], value); CHECKSTATUS(status);
-      }
-    }
-    else if (interp_method == LINEAR_INVDIST_LAND_TRACER){
-      if (grid->zdim==1){
-        status = spatial_interpolation_bilinear_invdist_land(xsi, eta, data2D[0], value); CHECKSTATUS(status);
-      }
-      else {
-        status = spatial_interpolation_trilinear_invdist_land(xsi, eta, zeta, data3D[0], value); CHECKSTATUS(status);
-      }
-    }
-    else {
-        return ERROR;    
-    }
-    return SUCCESS;
+    status = getCell3D(f, xi[igrid], yi[igrid], zi[igrid], ti[igrid], data3D, tii == 1); CHECKSTATUS(status);
+  }
+
+  // define a helper macro that will select the appropriate interpolation method
+  // depending on whether we need 2D or 3D
+#define INTERP(fn_2d, fn_3d)						\
+  do {									\
+    if (grid->zdim == 1) {						\
+      for (int i = 0; i < tii; i++) {					\
+	status = fn_2d(xsi, eta, data2D[i], &val[i]);			\
+	CHECKSTATUS(status);						\
+      }									\
+    } else {								\
+      for (int i = 0; i < tii; i++) {					\
+	status = fn_3d(xsi, eta, zeta, data3D[i], &val[i]);		\
+	CHECKSTATUS(status);						\
+      }									\
+    }									\
+  } while (0)
+
+  if ((interp_method == LINEAR) || (interp_method == CGRID_VELOCITY) ||
+      (interp_method == BGRID_VELOCITY) || (interp_method == BGRID_W_VELOCITY)) {
+    // adjust the normalised coordinate for flux-based interpolation methods
+    if ((interp_method == CGRID_VELOCITY) || (interp_method == BGRID_W_VELOCITY)) {
+      if (gridindexingtype == NEMO) {
+	// velocity is on the northeast of a tracer cell
+	xsi = 1;
+	eta = 1;
```

Comment:
Indentation not correct?

Reply:
Yep, my editor is being silly in C mode (clearly I don't use it much either!)